contracts/DataSourceDelegate.sol

```diff
@@ -1,86 +1,240 @@
+    if (_currentFundingCycle.reservedRate() != JBConstants.MAX_RESERVED_RATE)
+      return (_data.weight, _data.memo, IJBPayDelegate(address(0)));
+
+    // Get the token count that would be distributed by minting a new supply.
+    uint256 _mintTokenCountQuote = PRBMath.mulDiv(
+      _data.amount.value,
+      _currentFundingCycle.weight,
+      10**18
+    );
+
+    // Get the token count that would be distributed by swapping.
+    uint256 _swapTokenCountQuote = OracleLibrary.getQuoteAtTick(
+      OracleLibrary.consult(address(pool), uint32(twapPeriod)),
+      uint128(_data.amount.value),
+      address(weth),
+      address(jbx)
+    );
+    emit Test(_swapTokenCountQuote);
+    // Get the current overflow allowance.
+    (uint256 _currentAllowance, ) = IJBController(jbxTerminal.directory().controllerOf(projectId))
+      .overflowAllowanceOf(
+        _data.projectId,
+        _currentFundingCycle.configuration,
+        _data.terminal,
+        JBTokens.ETH
+      );
+
+    // Get the current used amount of overflow.
+    uint256 _usedAllowance = terminalStore.usedOverflowAllowanceOf(
+      IJBSingleTokenPaymentTerminal(address(_data.terminal)),
+      _data.projectId,
+      _currentFundingCycle.configuration
+    );
+
+    // If swapping yields a better rate and there's enough overflow allowance to cover the swap.
+    if (
+      _mintTokenCountQuote < _swapTokenCountQuote &&
+      _currentAllowance - _usedAllowance > _data.amount.value
+    ) {
+      // Get the weight which represents the amount being swapped for.
+      uint256 _swapWeight = PRBMath.mulDiv(_swapTokenCountQuote, 1E18, _data.amount.value);
+
+      // Get the weight which should be used to distribute reserved tokens given a 100% reserved rate.
+      uint256 _reservedWeight = PRBMath.mulDiv(
+        _swapWeight,
+        JBConstants.MAX_RESERVED_RATE,
+        JBConstants.MAX_RESERVED_RATE - reservedRate
+      ) - _swapWeight;
+
+      // Store the token count to swap. This will be referenced and reset in the delegate.
+      _swapTokenCount = _swapTokenCountQuote;
+
+      // Return the weight of reserved tokens to distribute, forward the memo, and set this contract as the delegate to execute the swap.
+      return (_reservedWeight, _data.memo, IJBPayDelegate(address(this)));
+    } else {
+      // Get the weight which should be used to distribute reserved tokens given a 100% reserved rate.
+      uint256 _reservedWeight = PRBMath.mulDiv(
+        _data.weight,
+        JBConstants.MAX_RESERVED_RATE,
+        JBConstants.MAX_RESERVED_RATE - reservedRate
+      ) - _data.weight;
+
+      // Store the token count to mint. This will be referenced and reset in the delegate.
+      _issueTokenCount = _mintTokenCountQuote;
+
+      // Get a reference to the weight of reserved tokens to distribute.
+      return (_reservedWeight, _data.memo, IJBPayDelegate(address(this)));
```

Comment:
I think this will not work if using a `_minReturnedTokens` in `pay(..)`
- with non-0 weight, terminal then tries to mint as next step, using the max reserved rate -> 0 token minted to beneficiary (ok as the delegate will do it as next step) but revert on `beneficiaryTokenCount < _minReturnedTokens`
- with 0 weight: well, same :D

Reply:
-> using _minReturnedTokens == 0